examples/jsm/renderers/nodes/core/NodeBuilder.js

```diff
@@ -0,0 +1,166 @@
+const VERSION = '1';
+
+class NodeBuilder {
+
+	constructor() {
+		
+		this.slots = { vertex: [], fragment: [] };
+		this.defines = { vertex: {}, fragment: {} };
+		
+		this.nodesData = new WeakMap();
+		
+		this.shader = undefined;
+		
+	}
+	
+	setMaterial( material ) {
+		
+		this.material = material;
+		
+	}
+	
+	addSlot( shader, slot ) {
+		
+		this.slots[ shader ].push( slot );
+		
+	}
+	
+	addDefine( shader, name, value ) {
+		
+		this.defines[ shader ][ name ] = value || '';
+		
+	}
+	
+	generateVec3( x, y, z ) {
+		
+		return `vec3( ${x}, ${y}, ${z} )`;
+		
+	}
+	
+	generateFloat( value ) {
+		
+		return value + ( value % 1 ? '' : '.0' );
+		
+	}
+	/*
+	createDataFromNode( node ) {
+		
+		return this.nodesData[ node ] = this.nodesData[ node ] || {};
+		
+	}
+	*/
+	createUniformFromNode( node ) {
+		
+		
+		
+	}
+	/*
+	analyzeNode( node ) {
+		
+		
+	}
+	*/
+	flowNode( node, output ) {
```

Comment:
Since `output` can be `undefined`, it's easier to read if `output` has a default value.
```js
flowNode( node, output === null ) {
```
We lately use `null` instead of `undefined` since it's more robust to test for `null` instead of relying on `undefined` values.